### PR TITLE
Fix hyperkube build for Azure

### PR DESCRIPTION
**Pull request by adelina-t**
cmd/hyperkube has been removed from kubernetes in 1.17 as such
we need a way to build hyperkube that takes that into account and
is also backwards compatible for jobs testing released k8s branches.

NOTE: This method will only work when --build=host-go

### CODE REVIEW

File: kubetest/azure.go [hunk 1]
```diff
@@ -719,21 +719,31 @@ func (c *Cluster) buildCcm() error {
...
 	if err := c.dockerLogin(); err != nil {
 		return err
 	}
 	log.Println("Building hyperkube.")
-	cmd := exec.Command("make", "-C", util.K8s("kubernetes"), "WHAT=cmd/hyperkube")
-	// dev-push-hyperkube will produce a lot of output to stdout. We should capture the output here.
-	cmd.Stdout = ioutil.Discard
-	if err := control.FinishRunning(cmd); err != nil {
-		return err
+	hyperkube_cmd_path := util.K8s("kubernetes/cmd/hyperkube")
```
Comment (chewong) [inline kubetest/azure.go]:
```suggestion
	hyperkubeCmdPath := util.K8s("kubernetes", "cmd", "hyperkube")
```

File: kubetest/azure.go [hunk 2]
```diff
@@ -719,21 +719,31 @@ func (c *Cluster) buildCcm() error {
...
+		cmd.Stdout = ioutil.Discard
+		if err := control.FinishRunning(cmd); err != nil {
+			return err
+		}
+		hyperkubeBin := util.K8s("kubernetes", "_output", "bin", "hyperkube")
+		push_cmd = exec.Command("make", "-C", util.K8s("kubernetes", "cluster", "images", "hyperkube"), "push", fmt.Sprintf("HYPERKUBE_BIN=%s", hyperkubeBin))
+
+	} else if os.IsNotExist(err) {
+		push_cmd = exec.Command("make", "-C", util.K8s("kubernetes", "cluster", "images", "hyperkube"), "push")
+
```
Comment (chewong) [inline kubetest/azure.go]: Nit: remove extra space.

File: kubetest/azure.go [hunk 3]
```diff
@@ -719,21 +719,31 @@ func (c *Cluster) buildCcm() error {
 
 func (c *Cluster) buildHyperKube() error {
 
+	var push_cmd *exec.Cmd
```
Comment (chewong) [inline kubetest/azure.go]: camelCase is preferred for local variables.

**Comment by chewong** [comment]
/hold

I don't think the new hyperkube image would work with customHyperkubeImage key in aks-engine API model. An `ENTRYPOINT` is added to hyperkube's Dockerfile, so commands like [this](https://github.com/Azure/aks-engine/blob/master/parts/k8s/cloud-init/artifacts/cse_install.sh#L233) won't work.

**Comment by dims** [comment]
@chewong you can add `--entrypoint` to override the entrypoint if you wish

**Comment by adelina-t** [comment]
@chewong Can we go ahead in the following fashion? : we merge this and continue testing with custom aks-engine build ( we have done so in the past and documented the reasons here when it was the case: https://github.com/kubernetes-sigs/windows-testing/blob/master/aks-engine-version-notes.md

And we fix this to use proper release builds when https://github.com/Azure/aks-engine/pull/2191 merges, as it seems it has a little more to go.

@PatrickLang What do you think?

**Comment by PatrickLang** [comment]
I think we should merge this. We haven't had test passes on Azure since 11/7. We need test passes working before we enter code freeze.

We can adjust the AKS-Engine release later, that's not a big deal.

**Comment by chewong** [comment]
@adelina-t @PatrickLang sounds good. I've opened a PR in aks-engine (https://github.com/Azure/aks-engine/pull/2298) which should fix the issue. Note that currently, https://github.com/Azure/aks-engine/pull/2191 will not honor `customHyperkubeImage` field in API models anymore. I had a discussion with @mboersma and we should continue supporting `customHyperkubeImage` for now to avoid breaking test-infra again. But in the long term, we are planning to move away from hyperkube and build individual components instead.

/hold cancel

**Comment by PatrickLang** [comment]
ok, let's get this going
/approve

**Comment by PatrickLang** [comment]
/assign @feiskyer 
Can you review and approve if this is still open when you get into the office?

**Comment by PatrickLang** [comment]
/assign @krzyzacy 
Are you able to approve?

**Pull request merged**